### DISCUSSION TITLE
How to extract info from IndirectObject?

### DISCUSSION THREAD

**Discussion by Nivdc** [py-pdf/pypdf#1394] [Q&A] [answered]
I am developing a tool to edit pdf bookmarks, the first thing I want to do is load existing bookmarks from the pdf file.
This should be easy, so I wrote:

```py 

from PyPDF2 import PdfReader

reader = PdfReader("test.pdf")

for idx,outline in enumerate(reader.outline):
    print(outline)

    if idx == 0:
        break

```

and then the output is:

```
{'/Title': 'aTitle', '/Page': IndirectObject(9, 0, 2640284515392), '/Type': '/FitH', '/Top': 622, '/Count': 467}
```

now, I want to know which page this bookmark reference to.
How can I do that?

<br />

Search for information about IndirectObject, some answers mention there is a method called ```getData()```.
But when I trying this method, the output is:
```
...
AttributeError: 'IndirectObject' object has no attribute 'getData'
```
So ... is this method has been removed?

<br />

By read [the source code of IndirectObject](https://github.com/py-pdf/PyPDF2/blob/main/PyPDF2/generic/_base.py).
I found there is a method called ```get_object()```.

I know that if I can get the PageObject this IndirectObject reference to, I can get the page number using the ```get_page_number(page)``` method.
But unfortunately ```get_object()``` method returns a:
```
<class 'PyPDF2.generic._data_structures.DictionaryObject'>
```
It contains the following:
```
{'/Type': '/Page', 
'/Contents': IndirectObject(8, 0, 2237557800000), 
'/MediaBox': [0, 0, 612, 792], 
'/Parent': IndirectObject(14, 0, 2237557800000), 
'/Resources': {'/ExtGState': {'/G3': IndirectObject(3, 0, 2237557800000)}, 
'/Font': {'/F4': IndirectObject(4, 0, 2237557800000), '/F5': IndirectObject(5, 0, 2237557800000), '/F6': IndirectObject(6, 0, 2237557800000), '/F7': IndirectObject(7, 0, 2237557800000)}, 
'/ProcSet': ['/PDF', '/Text', '/ImageB', '/ImageC', '/ImageI']}, 
'/StructParents': 0}
```

I didn't see anything like page numbers.

Is this DictionaryObject is the IndirectObject direct reference to? So there is no way to find out the PageObject I am looking for?
Or there is a PageObject, but ```get_object()``` skip it?
Or I am using the wrong way to find the Object referenced by the IndirectObject?

( ... whenever you feel like _It should be easy_ ... )

**Comment by pubpub-zz** [comment]
The Page Object within the structure does not contain a page index.
What you can do:
a) produce the 'flatten' list of PageObject :
`reader._flatten()`
This function is automatically called  when acessing to pages through `reader.pages[n]` property
once this table is generated to find the index of the page indirect object is quite easy:
`reader.flattened_pages.index(outline.raw_get("/Page"))`       <-- raw_get is here to prevent the automatic call to get_object()

**Reply by Nivdc** [reply]
Thanks for reply.

```reader.flattened_pages.index(outline.raw_get("/Page"))```
output:
```
...
ValueError: IndirectObject(2, 0, 2268570894192) is not in list
```

However, ```reader.flattened_pages.index(outline.page.get_object())``` works well, I can see that the indices start from 0.

**Reply by Nivdc** [reply]
I'm glad to find a solution to my problem.
But I still wonder what IndirectObject refers to, and how to find it out?

I would be more grateful if you could tell us more details.

**Reply by pubpub-zz** [reply]
IndirectObject are pointers

**Reply by Nivdc** [reply]
> IndirectObject are pointers

What is this pointer point to ?  
In this case, is it point to a DictionaryObject ?

How to dereference these pointers ? 
Just use ```get_object()``` ?

( Sorry for the bad description. I'm a little confused about the words point and refer. XD )

**Comment by pubpub-zz** [comment, marked as answer]
(Sorry for previous short answer, I had only my phone with me)
Objects (pages properties, contents,...) are stored within the PDF as objects which are identified with an id and a generation. those objects are stored in the pdf and most of the time the xref table stores the lookup table from (id,gen) to object position within the pdf. Within the pdf the objects will start with header (id) (gen) obj and a trailer endobj
the IndirectObject is a pointer which is encoded as "(id) (gen) R" and provide a way to point to objects. get_object() returns the pointed data.
as a simplifcation if you use the standard key call (such as ["/Page"]), an implicit call to the get_object() is provided.
note that when an IndirectObject are within an array (for example the list of annotations in a page) the "key"/index call does not call get_object and returns the IndirectObject (you have then to call get_object() by your own)

To complete, the get_object() is implemented on all objects but they will return the object themselves, so you don't need to check if it is an IndirectObject before calling get_object()

**Reply by Nivdc** [reply]
Thank you for your help, that's clear enough.
